src/workshops/gradients/gradientEditor.ts

```diff
@@ -2,54 +2,165 @@ import * as ko from "knockout";
 import template from "./gradientEditor.html";
 import { Component, Param, Event, OnMounted } from "@paperbits/common/ko/decorators";
 import { LinearGradientContract, LinearGradientColorStopContract, getLinearGradientString } from "../../contracts";
+import { LinearGradientViewModel, ColorStopViewModel } from "./linearGradientViewModel";
+import { Style, StyleSheet, StyleRule } from "@paperbits/common/styles";
 
 
 @Component({
     selector: "gradient-editor",
     template: template
 })
 export class GradientEditor {
-    public readonly direction: ko.Observable<number>;
-    public readonly colorStops: ko.ObservableArray<LinearGradientColorStopContract>;
     public readonly gradientPreview: ko.Observable<Object>;
+    public readonly gradientViewModel: ko.Observable<LinearGradientViewModel>;
+
+    public direction: ko.Observable<number>;
+    public dragging: boolean[];
+    public initialOffset: number;
+
+    @Param()
+    public readonly selectedGradient: ko.Observable<LinearGradientContract>;
+
+    @Event()
+    public readonly onSelect: (gradient: LinearGradientContract) => void;
 
     constructor() {
         this.initialize = this.initialize.bind(this);
-        this.addColorStop = this.addColorStop.bind(this);
 
-        this.direction = ko.observable(90);
-        this.colorStops = ko.observableArray();
-        this.colorStops.push({ color: "#ABDCFF", length: 0 });
-        this.colorStops.push({ color: "#0396FF", length: 100 });
-        this.gradientPreview = ko.observable();
+        this.gradientPreview = ko.observable<string>();
+        this.gradientViewModel = ko.observable();
+        this.selectedGradient = ko.observable();
+        this.direction = ko.observable<number>();
     }
 
     @OnMounted()
     public async initialize(): Promise<void> {
-        // const gradient = this.gradient();
+        this.gradientPreview(null);
+
+        this.gradientViewModel(this.selectedGradient ? 
+            new LinearGradientViewModel(this.selectedGradient()) 
+            : new LinearGradientViewModel(null));
+        this.direction(parseFloat(this.gradientViewModel().direction()));
+        this.direction.subscribe(deg => {
+            this.gradientViewModel().direction(deg + "deg");
+            this.applyChanges();
+            this.updateOnSelect();
+        });
+
+        this.configDragging();
+
+        this.attachFunction();
+        this.updateBackground();
+        
+        this.gradientViewModel().colorStops.subscribe(this.configDragging);
+    }
+
+    private configDragging(): void {
+        this.dragging = this.gradientViewModel().colorStops().map(() => false);
+    }
 
-        // const jss = await this.backgroundStylePlugin.contractToStyleRules(background);
-        // this.backgroundPreview({ backgroundPreview: jss });
+    public async attachFunction(): Promise<void> {
```

Comment:
I think subscribing for nested properties is not really needed here - you have just 2 sources of updates, adding colors and drag-end events, right?

Reply:
It will also change when the user modify the color or drag the length. I tried to subscribe the array only and it didn't update when I select the new color.